### ISSUE TITLE
[question] How to set rpath-link in conan recipe?

### ISSUE THREAD

**Issue by Tumb1eweed** [conan-io/conan#11531]
As I mention in title, I try to develop a conan recipe for TensorRT OSS + TensorRT GA， and  this thing happen:
```
[ 81%] Building CXX object parsers/onnx/third_party/onnx/CMakeFiles/onnx.dir/onnx/defs/nn/defs.cc.o
[ 81%] Linking CXX executable getSupportedAPITest
/usr/bin/ld: warning: libcublas.so.11, needed by /data/.data/tensorrt/8.2.5/_/_/source/TensorRT-8.4.1.5/lib/libnvinfer_plugin.so, not found (try using -rpath or -rpath-link)
/usr/bin/ld: warning: libcublasLt.so.11, needed by /data/.data/tensorrt/8.2.5/_/_/source/TensorRT-8.4.1.5/lib/libnvinfer_plugin.so, not found (try using -rpath or -rpath-link)
/usr/bin/ld: warning: libcudnn.so.8, needed by /data/.data/tensorrt/8.2.5/_/_/source/TensorRT-8.4.1.5/lib/libnvinfer_plugin.so, not found (try using -rpath or -rpath-link)
```
What's the proper way to set rpath-link or LD_LIBRARY_PATH when developing conan recipes?

**Comment by memsharded** [comment]
Hi @Tumb1eweed 

The ``LD_LIBRARY_PATH`` should be automatically set if you define the ``VirtualBuildEnv`` generator, it will create that file, containing the paths, that will be automatically injected before each ``self.run()`` call (that includes internals of ``CMake`` and other helpers)

**Comment by jcar87** [comment]
Hi @Tumb1eweed! Thanks for raising this issue.

The behaviour you are experiencing is described [here](https://stackoverflow.com/questions/24598047/why-does-ld-need-rpath-link-when-linking-an-executable-against-a-so-that-needs): you are creating an executable (`getSupportedAPITest`), linking it against a shared library (`libnvinfer_plugin.so`) and the GNU linker (`ld`) is trying to locate that library's indirect dependencies (`libcublas.so` and so on), but does not know where to find them. The linker will look in its default search path, so I am guessing in this case CUDA is not installed at the system level.

The solution for you will depend on whether or not you have the CUDA toolkit inside a Conan package or at the system level, as well as the CMakeLogic in the project you are building.

* If your CUDA toolkit is in a Conan package, and you depend on it to build TensorRT OSS - then @memsharded's suggestion should work - as the GNU linker will fall back on looking at `LD_LIBRARY_PATH` at link time, for indirect dependencies.

* If your CUDA toolkit was installed at the system level (somewhere in `/usr/local`) -  depending on the installation method, the default linker search path may have been overridden. I dont think this is the case for you, otherwise you wouldn't be getting that error.


If CMake has enough contextual information about the dependencies, it will set `-rpath` or `-rpath-link` for you as appropriate.
However, I can see that TensorRT OSS _disables_ this behaviour: https://github.com/NVIDIA/TensorRT/blob/main/CMakeLists.txt#L46

Relying on CMake (and for Conan for provide this information to CMake) would be preferred as it makes this more portable and less reliant on specific system installation.

In order to provide further assistance, could you specify:
* Whether the CUDA toolkit is installed via a Conan package, at the system level (`/usr/local`), or if it's in an arbitrary location
* Which Conan CMake integration you are using to build the TensorRT OSS recipe - either the legacy `conans.CMake`, or the newer `conan.tools.CMake` (which works in conjunction with the `CMakeDeps` and `CMakeToolchain` generators). - This last bit is relevant since I can see that TensorRT may have slightly different behaviours of a CMake toolchain file is passed.

Thanks!

**Comment by Tumb1eweed** [comment]
> Hi @Tumb1eweed! Thanks for raising this issue.
> 
> The behaviour you are experiencing is described [here](https://stackoverflow.com/questions/24598047/why-does-ld-need-rpath-link-when-linking-an-executable-against-a-so-that-needs): you are creating an executable (`getSupportedAPITest`), linking it against a shared library (`libnvinfer_plugin.so`) and the GNU linker (`ld`) is trying to locate that library's indirect dependencies (`libcublas.so` and so on), but does not know where to find them. The linker will look in its default search path, so I am guessing in this case CUDA is not installed at the system level.
> 
> The solution for you will depend on whether or not you have the CUDA toolkit inside a Conan package or at the system level, as well as the CMakeLogic in the project you are building.
> 
> * If your CUDA toolkit is in a Conan package, and you depend on it to build TensorRT OSS - then @memsharded's suggestion should work - as the GNU linker will fall back on looking at `LD_LIBRARY_PATH` at link time, for indirect dependencies.
> * If your CUDA toolkit was installed at the system level (somewhere in `/usr/local`) -  depending on the installation method, the default linker search path may have been overridden. I dont think this is the case for you, otherwise you wouldn't be getting that error.
> 
> If CMake has enough contextual information about the dependencies, it will set `-rpath` or `-rpath-link` for you as appropriate. However, I can see that TensorRT OSS _disables_ this behaviour: https://github.com/NVIDIA/TensorRT/blob/main/CMakeLists.txt#L46
> 
> Relying on CMake (and for Conan for provide this information to CMake) would be preferred as it makes this more portable and less reliant on specific system installation.
> 
> In order to provide further assistance, could you specify:
> 
> * Whether the CUDA toolkit is installed via a Conan package, at the system level (`/usr/local`), or if it's in an arbitrary location
> * Which Conan CMake integration you are using to build the TensorRT OSS recipe - either the legacy `conans.CMake`, or the newer `conan.tools.CMake` (which works in conjunction with the `CMakeDeps` and `CMakeToolchain` generators). - This last bit is relevant since I can see that TensorRT may have slightly different behaviours of a CMake toolchain file is passed.
> 
> Thanks!

Yes, my circumstance is the first case, I build tensorRT based on CUDA and CUDNN via Conan package.
And the second question, I use "from conans import ConanFile, CMake, tools", so conan.CMake it is
```    
   def build(self):
        env_build = RunEnvironment(self)
        with tools.environment_append(env_build.vars):
            cuda = self.dependencies["cuda"]
            cudnn = self.dependencies["cudnn"]
            self.run("LD_LIBRARY_PATH=${LD_LIBRARPY_PATH}:%s" % (cuda.package_folder + "/lib64:" + cudnn.package_folder + "/lib"))
            cmake = CMake(self)
            cmake.definitions["CUDAToolkit_ROOT"] = cuda.package_folder
            cmake.definitions["CUDNN_ROOT_DIR"] = cudnn.package_folder
            cmake.definitions["CUDA_VERSION"] = cuda.ref.version
            cmake.definitions["CUDNN_VERSION"] = cudnn.ref.version
            cmake.definitions["CMAKE_CUDA_COMPILER"] = cuda.package_folder + "/bin/nvcc"
            cmake.definitions["TRT_LIB_DIR"] = self.source_folder + "/TensorRT-8.4.1.5/lib"
            cmake.definitions["CMAKE_BUILD_TYPE"] = "Release"
            cmake.definitions["CMAKE_CXX_STANDARD"] = 14
            cmake.definitions["GPU_ARCHS"] = 75
            cmake.definitions["CMAKE_CUDA_ARCHITECTURES"] = 75
            cmake.definitions["TRT_OUT_DIR"] = self.build_folder
            cmake.configure()
            cmake.build()
            cmake.install()
```

**Comment by Tumb1eweed** [comment]
And I try VirtualbuildEnv like this, but it doesn't work...
```
    generators = "VirtualBuildEnv"
    ...
    def generate(this):
        vb = VirtualBuildEnv(self)
        vb.generate()
 ```

**Comment by Tumb1eweed** [comment]
> Hi @Tumb1eweed
> 
> The `LD_LIBRARY_PATH` should be automatically set if you define the `VirtualBuildEnv` generator, it will create that file, containing the paths, that will be automatically injected before each `self.run()` call (that includes internals of `CMake` and other helpers)


And I try VirtualbuildEnv like this, but it doesn't work...
```
    generators = "VirtualBuildEnv"
    ...
    def generate(this):
        vb = VirtualBuildEnv(self)
        vb.generate()
 ```

**Comment by Tumb1eweed** [comment]
> Hi @Tumb1eweed
> 
> The `LD_LIBRARY_PATH` should be automatically set if you define the `VirtualBuildEnv` generator, it will create that file, containing the paths, that will be automatically injected before each `self.run()` call (that includes internals of `CMake` and other helpers)

And I try virtualrunenv, still not useful

**Comment by memsharded** [comment]
Sorry this was not followed up back then.

I am closing now this ticket as outdated.

It is very likely that this is no longer an issue with new ``CMakeConfigDeps`` generator, generating real IMPORTED SHARED/STATIC targets, with the right properties. that defines -rpath-link correctly. Please try it, and open new tickets if there are any further issues or questions. Thanks for the feedback!
